### ISSUE TITLE
Upgrade to spaCy 3.x

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#274]
EDIT: Please refer to https://github.com/BlueBrain/Search/issues/274#issuecomment-801108330 for the updated scope.

---

## 🚀 Feature

- [ ] Add `spacy 3.x` to our dependencies.
- [ ] Re-run our DVC pipelines and report differences.
- [ ] Check compatibility with `prodigy`.
- [ ] Release a new version for `BlueBrain/Search`.

## Motivation

- [SpaCy released 3.x in 2021-01.](https://explosion.ai/blog/spacy-v3)
- It is not even clear how to access to old docs for 2.x on their website.

## Additional context
Add any other screenshot, context, or information about the problem here.

**Comment by EmilieDel** [comment]
PR #248 is starting the work to upgrade spaCy 3.x. 
- Dependencies have been changed (for `spaCy` and `scispaCy`)
- Make sure our tests are passing with the new version of `spaCy` and changed the API when needed. 

However, this issue is currently blocked:
- Prodigy did not release yet a stable version integrating `spaCy 3.x`. Currently, they have only a nightly program (check [link](https://support.prodi.gy/t/prodigy-nightly-spacy-v3-support-ui-for-overlapping-spans-improved-feeds-more/3861))

Once Prodigy has a stable version, actions still to be done:
- [ ] Re-run our DVC pipelines and report differences
- [ ] Check if the patch created by #268 is still needed (=check if latest `spaCy` version is including @jankrepl's commit)
- [ ] Release a new version for BlueBrain/Search

**Comment by pafonta** [comment]
Hello @EmilieDel,

It seems that one could use with spaCy 3 annotations made with Prodigy: https://spacy.io/usage/projects#prodigy.

This could unblock the work on PR #248. What do you think?

**Comment by pafonta** [comment]
Hello @FrancescoCasalegno and @EmilieDel,

This issue is about upgrading the code for spaCy 3.

What about opening one or two issues to:
1. migrate to the new feature of spaCy 3 named `projects`? ([projects on spaCy 3 doc](https://spacy.io/usage/projects))
2. migrate to the integration of spaCy 3 projects with DVC? ([DVC integration on spaCy 3 doc](https://spacy.io/usage/projects#dvc))

**Comment by FrancescoCasalegno** [comment]
@pafonta has a point here. The main thing to consider however is the potential implications that this change would have in terms of refactoring and changes in the documentation.

**Comment by EmilieDel** [comment]
> It seems that one could use with spaCy 3 annotations made with Prodigy: https://spacy.io/usage/projects#prodigy.

Hello @pafonta, 
This is a really good finding! However, in my opinion, this does not solve the entire transition to spacy 3.x. Indeed, to train our different models with the annotations, we use `prodigy.recipes` and their `train` function (see `data_and_models/pipelines/ner/train.py`). Either we wait for prodigy to release a stable version integrating spacy 3.x, either we need to reimplement the training ourselves. What do you think ? Am I missing something ?

**Comment by pafonta** [comment]
Hello @EmilieDel !

If we could afford to wait, it would be indeed better to wait.

Otherwise, there is the possibility to use today the annotations from Prodigy for spaCy 2 with spaCy 3.

However, this indeed implies to adopt `spacy train` ([doc](https://spacy.io/api/cli#train)) instead of `prodigy train` ([doc](https://prodi.gy/docs/recipes#train)). Unless we have access to Prodigy nightly, it is unclear at the moment if Prodigy for spaCy 3 will adopt the new spaCy projects feature (and consequently `spacy train`) or not. So, we don't know yet if the new version of Prodigy will have us to migrate to `spacy train` anyway.

Besides, one could argue that adopting `spacy train` instead of `prodigy train` is better:
- for an open source project like Blue Brain Search is now as Prodigy is not available without paying,
- for maintenability as it would remove the dependency to Prodigy while running the NER pipeline ([Dockerfile](data_and_models/pipelines/ner/Dockerfile)).

**Comment by pafonta** [comment]
Hello @EmilieDel and @FrancescoCasalegno !

Explosion has just provided a code example on how to go from the binary `.spacy` format to the spaCy v2 `JSON` format: https://support.prodi.gy/t/prodigy-nightly-spacy-v3-support-ui-for-overlapping-spans-improved-feeds-more/3861/7.

**Comment by FrancescoCasalegno** [comment]
After discussing with @pafonta , the scope of this ticket is refactored as follows.
The main reason is that we need to bring the code from a usable state to another usable state, while at the same time avoiding to add too many things to the scope of this ticket.

### Within Scope of #274 
- upgrade to `spacy 3.x`
- Add `preprocess.py` to get `.spacy` files from `.jsonl` annotations
- Add `config.cfg` to support `spacy train`
- Implement models training with `spacy train` and therefore get rid of `train.py`
- Prove that such a change does not significantly impact performance (results may be slightly different, but not much worse than models we had before!)

### Outside Scope of #274 
- Implement model evaluation using `spacy evaluate`. Issue #308 will address this.
- Start using `project.yml` and `project.lock`. At the moment they seem to duplicate most of the content of `dvc.yaml` and `dvc.lock` without bringing any significant advantage. Should things change on the `spaCy` side (e.g. great new features added) we may go back to this.

**Comment by pafonta** [comment]
After discussions with @FrancescoCasalegno, here are the refined goals for this issue and #248:

#### 1 - Move spaCy version requirement from `<3` to `>=3`.

This implies to:

- [x] Most of the work should already have been done in #248.
- [ ] Check which version of spaCy 3 includes the patch from #268.
- [ ] Pin in `setup.py` the corresponding version.
- [ ] Remove the patch from #268.

#### 2 - Convert annotations on-the-fly from `JSONL` to `.spacy`.

This implies to:

- [ ] Use `preprocess.py` from [here](https://github.com/explosion/projects/blob/v3/tutorials/ner_drugs/scripts/preprocess.py) to do the conversion.
- [ ] Don't forget to add licensing information in `preprocess.py` (MIT license).
- [ ] Add a corresponding DVC stage in `pipelines/ner/dvc.yaml`.

References:

- https://spacy.io/api/data-formats#binary-training

#### 3 - Migrate from `prodigy.recipes.train` to `spacy train`.

This implies to:

- [ ] Create a `config.cfg` in `pipelines/ner/`:
  - with `spacy init config config.cfg -l en -p ner -o accuracy -G`
  - and then replace the corresponding values from `prodigy_train_kwargs` in `pipelines/ner/params.yaml`
- [ ] Remove Prodigy from `pipelines/ner/Dockerfile`.
- [ ] Remove `pipelines/ner/train.py`.
- [ ] Run `dvc repro -f` in `pipelines/ner/`.

References:

- https://spacy.io/api/cli#train
- https://prodi.gy/docs/recipes#train

#### 4 - Validate changes

This implies to:

- [ ] Check changes (e.g. model performances).
- [ ] Validate with the team the changes.
- [ ] Run `dvc push`.
- [ ] Commit updated `pipelines/ner/dvc.lock`.
- [ ] Follow the traditional PR checklist (e.g. CI tests passing).

#### 5 - Release a new version of Blue Brain Search

This implies to:

- [ ] Add a new version tag.
- [ ] Put the new version section in `docs/source/whatsnew.rst`.

#### Notes

The new version Prodigy compatible with spaCy 3 will automatically create the `config.cfg` with `prodigy data-to-spacy`. To check the differences between the `config.cfg` created in this issue with the one of the new Prodigy version, a new issue has been created: #<FIXME>.

#### Out of scope

1. Migrating `add_er` to `spacy train` and `config.cfg`. See #<FIXME>.
2. Converting permanently annotations from `JSONL` to `.spacy` and removing the dedicated DVC stage. See #308.
3. Migrating to `spacy evaluate` and `Scorer`. See #308.
4. Adopting `project.yml` with `spacy project dvc` or keeping our `dvc.yaml`. See #<FIXME>.
